cmd/podman/common/completion.go

```diff
@@ -23,90 +23,75 @@ var (
 	LogLevels = []string{"debug", "info", "warn", "error", "fatal", "panic"}
 )
 
-func getContainers(toComplete string, statuses ...string) ([]string, cobra.ShellCompDirective) {
+type completeType int
+
+const (
+	// complete names and IDs after two chars
+	completeDefault completeType = iota
+	// only complete IDs
+	completeIDs
+	// only complete Names
+	completeNames
+)
+
+type keyValueCompletion map[string]func(s string) ([]string, cobra.ShellCompDirective)
+
+func getContainers(toComplete string, cType completeType, statuses ...string) ([]string, cobra.ShellCompDirective) {
 	suggestions := []string{}
 	listOpts := entities.ContainerListOptions{
 		Filters: make(map[string][]string),
 	}
 	listOpts.All = true
 	listOpts.Pod = true
+	if len(statuses) > 0 {
+		listOpts.Filters["status"] = statuses
+	}
 
-	// TODO: The api doesn't handle several different statuses correct see:
-	// https://github.com/containers/podman/issues/8344
-	// Instead of looping over the statuses we should be able to set
-	// listOpts.Filters["status"] = statuses
-
-	var containers []entities.ListContainer
-	var err error
-	if len(statuses) == 0 {
-		containers, err = registry.ContainerEngine().ContainerList(registry.GetContext(), listOpts)
-		if err != nil {
-			cobra.CompErrorln(err.Error())
-			return nil, cobra.ShellCompDirectiveError
-		}
-	} else {
-		for _, s := range statuses {
-			listOpts.Filters["status"] = []string{s}
-			res, err := registry.ContainerEngine().ContainerList(registry.GetContext(), listOpts)
-			if err != nil {
-				cobra.CompErrorln(err.Error())
-				return nil, cobra.ShellCompDirectiveError
-			}
-			containers = append(containers, res...)
-		}
+	containers, err := registry.ContainerEngine().ContainerList(registry.GetContext(), listOpts)
+	if err != nil {
+		cobra.CompErrorln(err.Error())
+		return nil, cobra.ShellCompDirectiveError
 	}
 
 	for _, c := range containers {
-		// include ids in suggestions if more then 2 chars are typed
-		if len(toComplete) > 1 && strings.HasPrefix(c.ID, toComplete) {
+		// include ids in suggestions if cType == completeIDs or
+		// more then 2 chars are typed and cType == completeDefault
+		if ((len(toComplete) > 1 && cType == completeDefault) ||
+			cType == completeIDs) && strings.HasPrefix(c.ID, toComplete) {
 			suggestions = append(suggestions, c.ID[0:12]+"\t"+c.PodName)
 		}
 		// include name in suggestions
-		if strings.HasPrefix(c.Names[0], toComplete) {
+		if cType != completeIDs && strings.HasPrefix(c.Names[0], toComplete) {
 			suggestions = append(suggestions, c.Names[0]+"\t"+c.PodName)
```

Comment:
I know this is not your addition, but what is the `"\t"+` intended to accomplish? On my system (zsh) it produces a very weird-looking completion list:
```console
$ ./bin/podman ps -a --filter id=<TAB>
id=276af9483c22  id=a722b8b4c064  id=5aa479ac915d  id=dc7cecab1356
id=3821798b331e  id=e00da78572a9  id=ed308f016b78  id=1a2b7290886c
id=c5e5427ef032  id=413e9b86c8d0                                    -- mypod      <<< this is weird
id=145f126fd181                                                                id=7b254fe8286b     <<< so is the extra spacing
```
Similarly with `--name=<TAB>`. If I remove the tab part and recompile, I get a (normal, expected) list of just IDs.

Reply:
That's the description part. It will show you the pod name if the container is in a pod. I don't use zsh but with bash it looks fine:
```
$ bin/podman ps --filter id=
0da6096d149d                 57e08662abe8                 936761df281d                 bfe79af8a30d                 d2e3bc52e2bf                 
355bcc085321  (test)         5b408e8c01e0                 a73f295f2f73  (angry_allen)  c6cc999b6c71 
```

Reply:
How does it look if you do `podman rm [TAB]`?

Reply:
That's my zsh output:
```
% bin/podman ps --filter id=
id=355bcc085321  -- test
id=a73f295f2f73  -- angry_allen
id=0da6096d149d  id=57e08662abe8  id=5b408e8c01e0  id=936761df281d  id=bfe79af8a30d  id=c6cc999b6c71  id=d2e3bc52e2bf
```
The formatting works ok for me although I think bash and also fish are looking better.

Reply:
You can also globally turn off descriptions for podman if you want.

Reply:
I don't get those extra bits (parenthesized pod mane) in bash, only CIDs; there must be a bash option that I'm not setting. `podman rm <TAB>` gives me similar-looking results:
```
$ ./bin/podman rm <TAB>
2d9d3c73021a-infra  gifted_dhawan
mystifying_neumann  unruffled_hypatia  dreamy_kowalevski  keen_perlman
xenodochial_ellis   reverent_cannon    tender_dijkstra    kind_hellman  -- mypod
mystifying_buck                                                                romantic_burnell
```
FWIW **all** of those containers were run with `--pod mypod`. This may be an upstream issue in completion internals.

Reply:
>That's my zsh output:

That looks good, it's what I tend to expect with zsh additional-comment info. OK, I'll assume it's something broken on my end, and will not pursue it further for now. Thanks for the confirmation that it looks good on your end.

Reply:
OK I see zsh groups the suggestion with the same description. That's useful:
```
% bin/podman ps --filter id=0da6096d149d
id=355bcc085321  id=82939027f0ef  id=45a4a1acd5b6  -- test                                                                                                                                   
id=a73f295f2f73                                    -- angry_allen                                                                                                                            
id=0da6096d149d            id=57e08662abe8            id=5b408e8c01e0            id=936761df281d            id=bfe79af8a30d            id=c6cc999b6c71            id=d2e3bc52e2bf
```

Reply:
@edsantiago I have turned off descriptions in bash by default since they are natively supported. You would need to run `source <(bin/podman completion bash)` to test.